Use Aspect Ratio Computed From Presentational Hints Before Video Loads.

sammygill opened the pull request:
#### bda70c08fc70def99434a3fb9bc8c8ff43fea239
<pre>
Use Aspect Ratio Computed From Presentational Hints Before Video Loads.
<a href="https://bugs.webkit.org/show_bug.cgi?id=245001">https://bugs.webkit.org/show_bug.cgi?id=245001</a>
rdar://99755477

Reviewed by Brent Fulgham.

When the width and height attributes are provided to video elements,
those values should contribute to the aspect-ratio property with the
form: auto width x height. This is the aspect-ratio that should be used
when sizing the video before its metadata have been received. Once the
metadata has been received, we should instead use the aspect-ratio
provided by the intrinsic sizes. Another subtle trait about the video
element is that its default object size (300px x 150px) should not
contribute to the aspect-ratio at any point in time.

To help match this behavior, the RenderBox::boxAspectRatio virtual
method was created to help determine the appropriate aspect ratio for
different boxes. The default behavior is to just return
style().logicalAspectRatio(), so that the behavior can remain the same
in RenderBox::computeMinMaxLogicalHeightFromAspectRatio and
RenderBox::computeMinMaxLogicalWidthFromAspectRatio. In scenarios where
this may not be correct, such as with RenderVideo, the subclass can
override the behavior. For RenderVideo&apos;s case, we need to determine
whether or not we have received the video metadata. If we have, we
should return the aspect-ratio provided by the intrinsic size of the
video. Otherwise, we can use the aspect ratio provided by either the
CSS property or the presentational hints.

* LayoutTests/imported/w3c/web-platform-tests/html/rendering/replaced-elements/attributes-for-embedded-content-and-images/video-aspect-ratio-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/replaced-elements/attributes-for-embedded-content-and-images/video-aspect-ratio.html:
Imported most recent WPT test

* LayoutTests/media/video-intrinsic-width-height.html:
The final video element in this test should have a final dimension of
300px x 300px. This is because:

    1.) Since there is not actually any video content, this box will not
        have any intrinsic width/height.
    2.) There is an inline style that overrides the width/height
        attributes, so it will have width and height of auto.
    3.) Even though the width/height style is overridden, the attributes
        still contribute to the aspect ratio, giving it an aspect-ratio
        value of &quot;auto 100/100&quot;
    4.) When computing the automatic width we fall back to the default
        object size value, which is 300px.
    5.) To compute the height we take the aspect-ratio we computed and
        arrive at a height of 300px.

* Source/WebCore/rendering/RenderReplaced.cpp:
(WebCore::RenderReplaced::computeAspectRatioInformationForRenderBox const):
The new code path provides the width and height based on the aspect
ratio for video elements. Even though it seems like this logic should
work for any ReplacedElement, it seems to fail various tests. For now
we may want to just keep this logic separated for RenderVideo objects
specifically.

(WebCore::RenderReplaced::computeIntrinsicRatioInformation const):
We need to make sure we do not override the computed aspect ratio with
the values of the default object size for video elements. This is
because the default object size is not an intrinsic size and should not
contribute to the aspect ratio. Basically, we need to check out if we
have an intrinsic size set and bail out early if we do.
RenderVideo::calculateIntrinsicSize contains logic that decides if the
size needs to be set to the default size, so we can base most off of the
logic off of that.

(WebCore::isVideoWithDefaultObjectSize):
* Source/WebCore/rendering/RenderReplaced.h:
* Source/WebCore/rendering/RenderVideo.cpp:
(WebCore::RenderVideo::calculateIntrinsicSize):
(WebCore::RenderVideo::computeReplacedLogicalWidth const):
(WebCore::RenderVideo::hasVideoMetadata const):
(WebCore::RenderVideo::hasPosterFrameSize const):
(WebCore::RenderVideo::hasDefaultObjectSize const):
* Source/WebCore/rendering/RenderVideo.h:

Canonical link: <a href="https://commits.webkit.org/255743@main">https://commits.webkit.org/255743@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/43f77f412272e7ea9f91a7375a9cf156004e484a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/93421 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/2617 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/24065 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/103093 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/163426 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/97419 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/2627 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/30919 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/85787 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/99198 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/99084 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/1849 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/79872 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/28786 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/83762 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/83511 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/71849 "Found 3 new API test failures: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/event-listener, /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/state, /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/state-changed (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/37302 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/17378 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/35132 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/18628 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3974 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/39005 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/41155 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/40941 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/37851 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

cathiechen commented on Source/WebCore/rendering/RenderVideo.cpp:
```diff
@@ -366,6 +367,18 @@ bool RenderVideo::foregroundIsKnownToBeOpaqueInRect(const LayoutRect& localRect,
+double RenderVideo::boxAspectRatio() const
+{
+    if (!shouldApplySizeContainment() && hasVideoMetadata())
+        return computeIntrinsicAspectRatio();
```
I'm not sure if this is needed, for `aspect-ratio: auto w / h` makes this happen, if there is natural aspect-ratio, then use natural aspect-ratio instead of `w / h`.
According to the discussion in https://github.com/w3c/csswg-drafts/issues/7524#issuecomment-1204462924. I think maybe we should start from distinguish "the default object size", and make sure it does not affect aspect-ratio?

sammygill replied:
Yeah you are right this method is not needed. Went ahead and removed it

cathiechen commented on Source/WebCore/rendering/RenderReplaced.cpp:
```diff
@@ -492,6 +501,11 @@ void RenderReplaced::computeIntrinsicRatioInformation(FloatSize& intrinsicSize,
     }
+
+    const RenderVideo* videoBox;
+    if (is<RenderVideo>(this) && (videoBox = downcast<RenderVideo>(this)) && !videoBox->hasVideoMetadata() && style().hasAspectRatio()) 
```
Can we combine this to the previous `if (style().hasAspectRatio())`? And according to `RenderVideo::calculateIntrinsicSize`, to distinguish if intrinsicSize is from the default size or not, looks like not only we need to check `hasVideoMetadata`, but also `shouldApplySizeContainment` and the poster image.

sammygill replied:
Modified the logic and added those other checks. Let me know what you think or if you see any issues

cathiechen commented on Source/WebCore/rendering/RenderReplaced.cpp:
```diff
@@ -425,10 +427,17 @@ void RenderReplaced::computeAspectRatioInformationForRenderBox(RenderBox* conten
-        // can only be fixed once subpixel precision is available for things like intrinsicWidth/Height - which include zoom!
-        constrainedSize.setWidth(RenderBox::computeReplacedLogicalHeight() * intrinsicSize.width() / intrinsicSize.height());
-        constrainedSize.setHeight(RenderBox::computeReplacedLogicalWidth() * intrinsicSize.height() / intrinsicSize.width());
+        if (is<RenderVideo>(this)) {
```
If we can distinguish that intrinsicSize is from the default size, here we can set `width = RenderBox::computeReplacedLogicalWidth();`, and use `intrinsicRatio` to calculate height.

sammygill replied:
Sorry, with the changes I'm not sure about this one anymore. Do you think this is still an issue with the changes?

`computeIntrinsicRatioInformation` should now should compute `intrinsicRatio` correctly and not allow the default object size to influence it. If the object is a `RenderVideo`, then we can do what you said and use `intrinsicRatio`

cathiechen replied:
Yeah, looks fine now. However, I think here it would be more accurate to use `hasDefaultObjectSize` same to `computeIntrinsicRatioInformation`. Maybe we could move `hasDefaultObjectSize` to RenderVideo?

The other part looks good to me:)

sammygill replied:
Ok yeah that would definitely make more sense. Added hasDefaultObjectSize to RenderVideo. I also modified the casting to RenderVideo to use dynamicCast so that it is more concise. How does it look now?

cathiechen replied:
I guess it should be `if (is<RenderVideo>(*this) && downcast<RenderVideo>(*this).hasDefaultObjectSize())` here?

sammygill replied:
Hmmm wouldn't we want to have the code take this new path if we want to compute the video's constrained size using the aspect ratio? We would compute the width and then use the aspect ratio with it to compute the height. Does that sound right?

cathiechen replied:
Hmm, I think only video using the default size takes this path. Regarding other video, is there specification supported this change?

sammygill replied:
Yeah I looked over it again and it looks like that change is fine and I went ahead and made it. What I guess I'm not sure about is how the code below in the else branch (the original code) is correct. I think I may be missing some sort of context, but it doesn't seem right to be computing both the width and height like that using the aspect ratio. That may be a conversation to have at another time though.....

Let me know what you think of the changes

cathiechen replied:
OK, seems fine. My original thought is that this patch only changes the behavior of video using the default size, that's why I think here it should add `downcast<RenderVideo>(*this).hasDefaultObjectSize()`. I think the `if` path looks good to me. As to the `else` path, I'm not very familiar with it, the last modification is in 2012.

cathiechen commented on Source/WebCore/rendering/RenderReplaced.cpp:
```diff
@@ -45,10 +45,12 @@
 #include "RenderedDocumentMarker.h"
 #include "Settings.h"
 #include "VisiblePosition.h"
+#include "rendering/RenderVideo.h"
```
Looks like this line is not needed.

cathiechen commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -5466,13 +5466,14 @@ std::pair<LayoutUnit, LayoutUnit> RenderBox::computeMinMaxLogicalWidthFromAspect
     if (style().logicalMinHeight().isSpecified()) {
         if (LayoutUnit blockMinSize = constrainLogicalHeightByMinMax(LayoutUnit(), std::nullopt); blockMinSize > LayoutUnit())
-            transferredMinSize = inlineSizeFromAspectRatio(horizontalBorderAndPaddingExtent(), verticalBorderAndPaddingExtent(), style().logicalAspectRatio(), style().boxSizingForAspectRatio(), blockMinSize);
+            transferredMinSize = inlineSizeFromAspectRatio(horizontalBorderAndPaddingExtent(), verticalBorderAndPaddingExtent(), aspectRatio, style().boxSizingForAspectRatio(), blockMinSize);
```
Nice catch! There might be a problem here if the value of aspect-ratio is `auto w / h`. We are not sure which value is used, the natural aspect-ratio or `w / h`. But I don't think we need `boxAspectRatio` here, I think we need to find out which value is used here instead.

sammygill replied:
Yeah so that is actually why I added boxAspectRatio. For most objects, style().logicalAspectRatio() would be fine, but for RenderVideo it depends on whether we have the natural aspect-ratio or not. RenderView overrides it and uses hasVideoMetadata() to distinguish whether or not we have the default object size since the default object size is used before the metadata is received.

Instead of having boxAspectRatio, should we instead move the logic here to determine which aspect ratio to use like in RenderVideo::boxAspectRatio?

cathiechen replied:
Hmm, double-checked, the specification and the code. It's fine to use `style().logicalAspectRatio()` here.
First, the value of `style().logicalAspectRatio()` is always `w / h`, it does not consider `aspect-ratio: auto w / h` scenario.
Then, per [1], the value auto && [<ratio>](https://www.w3.org/TR/css-values-4/#ratio-value) only affects the replaced elements. And it looks like `RenderBox::computeMinMaxLogicalWidthFromAspectRatio` does not handle replaced elements at all.
So it seems fine to use `style().logicalAspectRatio()` here. `auto w / h` should be same to `w / h` for non-replaced elements. Do you agree with me here?

As to the replaced element, like img and video:
It is handled in `RenderReplaced::computeReplacedLogicalWidth`, looks like it does not consider the transferred min/max sizes at all. Maybe we need a patch to handle it here?  Maybe in `RenderBox::computeReplacedLogicalWidthRespectingMinMaxWidth`?

As to RenderVideo:
Yes, `hasVideoMetadata()` could distinguish it is using the default size or not. However, I think it might be nature to handle it in `RenderReplaced::computeAspectRatioInformationForRenderBox`, i.e, don't let "the default size" affect `intrinsicRatio`.

[1] https://www.w3.org/TR/css-sizing-4/#preferred-aspect-ratio

sammygill replied:
Yes you are right about the code in RenderBox. I removed the `boxAspectRatio `method so it is calling `style().logicalAspectRatio()` again. 

Since `RenderReplaced::computeReplacedLogicalWidth` is called by `RenderVideo::computeReplacedLogicalWidth`, I added a call to `computeReplacedLogicalWidthRespectingMinMaxWidth` inside of the RenderVideo code which should solve that issue I think.

For RenderVideo, I added a comment to the additional one you had here: https://github.com/WebKit/WebKit/pull/4509#discussion_r983852491

cathiechen reviewed (commented):
Hi, sorry for the slow reply!
I checked the discussion here https://github.com/w3c/csswg-drafts/issues/7524.
It looks like the conclusion is "Clarify that aspect-ratio:auto doesn't pull aspect ratio from the default object size, it can only pull from an actual loaded resource". So my comments are based on this conclusion.

brentfulgham commented on Source/WebCore/rendering/RenderReplaced.h:
```diff
@@ -52,14 +52,14 @@ class RenderReplaced : public RenderBox {
 
     double computeIntrinsicAspectRatio() const;
 
+    void computeIntrinsicRatioInformation(FloatSize& intrinsicSize, double& intrinsicRatio) const override;
```
Nit: Should have blank line after this method declaration.

sammygill replied:
Changed

brentfulgham commented on Source/WebCore/rendering/RenderReplaced.cpp:
```diff
@@ -425,10 +426,17 @@ void RenderReplaced::computeAspectRatioInformationForRenderBox(RenderBox* conten
-        // can only be fixed once subpixel precision is available for things like intrinsicWidth/Height - which include zoom!
-        constrainedSize.setWidth(RenderBox::computeReplacedLogicalHeight() * intrinsicSize.width() / intrinsicSize.height());
-        constrainedSize.setHeight(RenderBox::computeReplacedLogicalWidth() * intrinsicSize.height() / intrinsicSize.width());
+        if (auto* video = dynamicDowncast<RenderVideo>(this); video && video->hasDefaultObjectSize()) {
```
Nit: Since you don't use `video` after this line, it might be clearer to have a little helper function:

```
static bool isVideoWithDefaultObjectSize(const RenderReplaced* maybeVideo)
{
   if (auto* video = dynamicDowncast<RenderVideo>(maybeVideo))
       return video->hasDefaultObjectSize();
   return false;
}
```

Then the conditional would just be:

```
if (isVideoWithDefaultObjectSize(this)) {
    ....
} else {
    ....
}
```

sammygill replied:
That is definitely easier to read! Changed

brentfulgham commented on Source/WebCore/rendering/RenderReplaced.cpp:
```diff
@@ -489,7 +497,7 @@ void RenderReplaced::computeIntrinsicRatioInformation(FloatSize& intrinsicSize,
     if (style().hasAspectRatio()) {
         intrinsicRatio = style().logicalAspectRatio();
-        if (style().aspectRatioType() == AspectRatioType::Ratio)
+        if (auto* video = dynamicDowncast<RenderVideo>(this); style().aspectRatioType() == AspectRatioType::Ratio || (video && video->hasDefaultObjectSize()))
```
Then you could use that helper function here:

```
if (style().aspectRatioType() == AspectRatioType::Ratio || isVideoWithDefaultObjectSize(this)) {
    ....
}
```

sammygill replied:
Changed

brentfulgham reviewed (approved):
r=me, but please consider my suggestions for making the code a tiny bit more readable.

sammygill committed:
Use Aspect Ratio Computed From Presentational Hints Before Video Loads.
https://bugs.webkit.org/show_bug.cgi?id=245001
rdar://99755477

Reviewed by Brent Fulgham.

When the width and height attributes are provided to video elements,
those values should contribute to the aspect-ratio property with the
form: auto width x height. This is the aspect-ratio that should be used
when sizing the video before its metadata have been received. Once the
metadata has been received, we should instead use the aspect-ratio
provided by the intrinsic sizes. Another subtle trait about the video
element is that its default object size (300px x 150px) should not
contribute to the aspect-ratio at any point in time.

To help match this behavior, the RenderBox::boxAspectRatio virtual
method was created to help determine the appropriate aspect ratio for
different boxes. The default behavior is to just return
style().logicalAspectRatio(), so that the behavior can remain the same
in RenderBox::computeMinMaxLogicalHeightFromAspectRatio and
RenderBox::computeMinMaxLogicalWidthFromAspectRatio. In scenarios where
this may not be correct, such as with RenderVideo, the subclass can
override the behavior. For RenderVideo's case, we need to determine
whether or not we have received the video metadata. If we have, we
should return the aspect-ratio provided by the intrinsic size of the
video. Otherwise, we can use the aspect ratio provided by either the
CSS property or the presentational hints.

* LayoutTests/imported/w3c/web-platform-tests/html/rendering/replaced-elements/attributes-for-embedded-content-and-images/video-aspect-ratio-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/replaced-elements/attributes-for-embedded-content-and-images/video-aspect-ratio.html:
Imported most recent WPT test

* LayoutTests/media/video-intrinsic-width-height.html:
The final video element in this test should have a final dimension of
300px x 300px. This is because:

    1.) Since there is not actually any video content, this box will not
        have any intrinsic width/height.
    2.) There is an inline style that overrides the width/height
        attributes, so it will have width and height of auto.
    3.) Even though the width/height style is overridden, the attributes
        still contribute to the aspect ratio, giving it an aspect-ratio
        value of "auto 100/100"
    4.) When computing the automatic width we fall back to the default
        object size value, which is 300px.
    5.) To compute the height we take the aspect-ratio we computed and
        arrive at a height of 300px.

* Source/WebCore/rendering/RenderReplaced.cpp:
(WebCore::RenderReplaced::computeAspectRatioInformationForRenderBox const):
The new code path provides the width and height based on the aspect
ratio for video elements. Even though it seems like this logic should
work for any ReplacedElement, it seems to fail various tests. For now
we may want to just keep this logic separated for RenderVideo objects
specifically.

(WebCore::RenderReplaced::computeIntrinsicRatioInformation const):
We need to make sure we do not override the computed aspect ratio with
the values of the default object size for video elements. This is
because the default object size is not an intrinsic size and should not
contribute to the aspect ratio. Basically, we need to check out if we
have an intrinsic size set and bail out early if we do.
RenderVideo::calculateIntrinsicSize contains logic that decides if the
size needs to be set to the default size, so we can base most off of the
logic off of that.

(WebCore::isVideoWithDefaultObjectSize):
* Source/WebCore/rendering/RenderReplaced.h:
* Source/WebCore/rendering/RenderVideo.cpp:
(WebCore::RenderVideo::calculateIntrinsicSize):
(WebCore::RenderVideo::computeReplacedLogicalWidth const):
(WebCore::RenderVideo::hasVideoMetadata const):
(WebCore::RenderVideo::hasPosterFrameSize const):
(WebCore::RenderVideo::hasDefaultObjectSize const):
* Source/WebCore/rendering/RenderVideo.h:

Canonical link: https://commits.webkit.org/255743@main

Loirooriol commented on LayoutTests/imported/w3c/web-platform-tests/html/rendering/replaced-elements/attributes-for-embedded-content-and-images/video-intrinsic-width-height.html:
```diff
@@ -21,7 +25,7 @@
              verify that width/height does not influence intrinsic ratio -->
         <video title="both width/height attributes and style"
-               data-expected-width="300" data-expected-height="150"
+               data-expected-width="300" data-expected-height="300"
```
All browsers say 150, I think I'm going to revert this.
https://wpt.fyi/results/html/rendering/replaced-elements/attributes-for-embedded-content-and-images/video-intrinsic-width-height.html

Loirooriol replied:
Ah, no, Firefox fails but isn't 150. I guess that's https://github.com/w3c/csswg-drafts/issues/10605

Loirooriol replied:
Ah, you already reverted in https://github.com/WebKit/WebKit/pull/10386, but didn't export to WPT.
